WebDAV Fix current Android / iOS clients

lab-at-nohl opened the pull request:
I revised my PR from 3 month ago and aligned to today's app versions. Tested with **Android app release 4.3.0** and **iPhone app version 12.3.1**. While Android app is very robust/flexible in the iPhone app a lot of stuff is hardcoded (service discovery for SSO by webfinger, xml namespaces, http header for GET requests). 

Please merge before the yearly release - I need it ;-)

pschuele commented:
hi Johannes,
I ran our complete test-suite on your commit and got some errors, which need to be fixed before we can merge that:

~~~
There were 2 failures:
1) Tinebase_WebDav_Plugin_OwnCloudTest::testGetPropertiesForSharedDirectoryRights
Failed asserting that '<?xml version="1.0"?>\n
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:owncloud="http://owncloud.org/ns"><d:response><d:href>/remote.php/dav/files/tine20admin/</d:href><d:propstat><d:prop><owncloud:getlastmodified/><owncloud:getcontentlength/><owncloud:resourcetype/><owncloud:permissions/><owncloud:checksums/><owncloud:share-types/><owncloud:data-fingerprint/><owncloud:getetag/><owncloud:id/></d:prop><d:status>HTTP/1.1 404 Not Found</d:status></d:propstat></d:response><d:response><d:href>/remote.php/dav/files/tine20admin/shared/</d:href><d:propstat><d:prop><owncloud:getlastmodified/><owncloud:getcontentlength/><owncloud:resourcetype/><owncloud:permissions/><owncloud:checksums/><owncloud:share-types/><owncloud:data-fingerprint/><owncloud:getetag/><owncloud:id/></d:prop><d:status>HTTP/1.1 404 Not Found</d:status></d:propstat></d:response><d:response><d:href>/remote.php/dav/files/tine20admin/Admin%2c%20tine%20%c2%ae/</d:href><d:propstat><d:prop><owncloud:getlastmodified/><owncloud:getcontentlength/><owncloud:resourcetype/><owncloud:permissions/><owncloud:checksums/><owncloud:share-types/><owncloud:data-fingerprint/><owncloud:getetag/><owncloud:id/></d:prop><d:status>HTTP/1.1 404 Not Found</d:status></d:propstat></d:response></d:multistatus>\n
' contains "<owncloud:permissions>SCK</owncloud:permissions>".
/builds/tine20/tine20/tests/tine20/Tinebase/WebDav/Plugin/OwnCloudTest.php:311
2) Tinebase_WebDav_Plugin_OwnCloudTest::testInvalidOwnCloudVersion
Failed asserting that 207 is identical to 500.
/builds/tine20/tine20/tests/tine20/Tinebase/WebDav/Plugin/OwnCloudTest.php:181
/builds/tine20/tine20/tests/tine20/Tinebase/WebDav/Plugin/OwnCloudTest.php:352

There were 8 errors:
1) Tinebase_Server_WebDAVTests::testServer
PHPUnit\Framework\Exception: PHP Deprecated:  preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated in /builds/tine20/tine20/tine20/Tinebase/WebDav/Plugin/OwnCloud.php on line 189
2) Tinebase_Server_WebDAVTests::testServerWithAuthorizationHeader
PHPUnit\Framework\Exception: PHP Deprecated:  preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated in /builds/tine20/tine20/tine20/Tinebase/WebDav/Plugin/OwnCloud.php on line 189
3) Tinebase_Server_WebDAVTests::testServerWithAuthorizationEnv
PHPUnit\Framework\Exception: PHP Deprecated:  preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated in /builds/tine20/tine20/tine20/Tinebase/WebDav/Plugin/OwnCloud.php on line 189
4) Tinebase_Server_WebDAVTests::testServerWithAuthorizationRemoteUser
PHPUnit\Framework\Exception: PHP Deprecated:  preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated in /builds/tine20/tine20/tine20/Tinebase/WebDav/Plugin/OwnCloud.php on line 189
5) Tinebase_Server_WebDAVTests::testPropfindCurrentUserPrincipal
PHPUnit\Framework\Exception: PHP Deprecated:  preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated in /builds/tine20/tine20/tine20/Tinebase/WebDav/Plugin/OwnCloud.php on line 189
6) Tinebase_Server_WebDAVTests::testPropfindPrincipal
PHPUnit\Framework\Exception: PHP Deprecated:  preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated in /builds/tine20/tine20/tine20/Tinebase/WebDav/Plugin/OwnCloud.php on line 189
7) Tinebase_Server_WebDAVTests::testReportQuery
PHPUnit\Framework\Exception: PHP Deprecated:  preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated in /builds/tine20/tine20/tine20/Tinebase/WebDav/Plugin/OwnCloud.php on line 189
8) Tinebase_Server_WebDAVTests::testPropfindThundebird
PHPUnit\Framework\Exception: PHP Deprecated:  preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated in /builds/tine20/tine20/tine20/Tinebase/WebDav/Plugin/OwnCloud.php on line 189
ERRORS!
~~~

please send me a note when you need help resolving the problems.

lab-at-nohl commented:
As I haven't touched WebDAV permissions, failure 1) maybe related to a patch from @ccheng-dev two days ago?
https://github.com/tine-groupware/tine/commit/0cad63d4aa8b1d18716c49c1824d7a175a967db4

I will check for failure 2) and the deprecation errors. 

Btw. can I execute a single test?

Thanks!

lab-at-nohl commented:
Hello, @pschuele and @ccheng-dev. Despite my first guess, **Permissions** are fine. I had changed Laminas Request Headers to $_SERVER for *User-Agent* because in the parent classes this was widely used; that broke the tests. 

I've redone the *User-Agent* detection. That solved failure 1) and half of failure 2). However, I had to dive into phpunit a little, I don't think `testInvalidOwnCloudVersion` has ever worked. Changing http headers is not possible after `self::setUp()` -- I fixed that, too. If I find time I will write some tests for the iPhone client. 

My limited run of phpunit now reads:
`OK (12 tests, 46 assertions)`

paulmhh commented on tests/tine20/Tinebase/WebDav/Plugin/OwnCloudTest.php:
this needs investigation, I dont see why this is necessary, it shouldn't be merged like that.

it is perfectly possible to use Tinebase_Core::set(Tinebase_Core::REQUEST [or alike] [not 'request'], $request);

lab-at-nohl replied:
Do you mean line 351 `Tinebase_Core::set('request', $request);`? This is what I have found, I didn't invent it. Anyways, with my change it is removed...

The basic problem is that a second use of `Tinebase_Core::set(Tinebase_Core::REQUEST, $request);` inside a test suite **does not overwrite its first declaration in the inherited function `::setUp()`**. I double checked; the User Agent will not be overwritten (thus, the test does not work). That's why I assume that `public function testInvalidOwnCloudVersion()` never has worked before.

lab-at-nohl replied:
I had tried this. Did this once again with the original files for double checking. Again, the Request from ::setUp() is reused. Hence, the User Agent is not changed...

/app/tine20/Tinebase/WebDav/Plugin/OwnCloud.php with Output for Testing
``` php
    public function initialize(\Sabre\DAV\Server $server)
    {
        $this->server = $server;
        // Although parent class heavily uses $_SERVER, the test suite requires $request...
        $request = Tinebase_Core::get(Tinebase_Core::REQUEST); 
        $this->useragent = isset($request) ? $request->getHeader('user-agent')->toString() : $_SERVER['HTTP_USER_AGENT'] ?? null;
        $server->on('propFind', array($this, 'propFind'));
        echo ">>>>>>>>>>>" . $this->useragent . "<<<<<<<<<";
```

/app/tests/tine20/Tinebase/WebDav/Plugin/OwnCloudTest.php:
``` php
    public function testInvalidOwnCloudVersion()
    {
        // use old owncloud user agent!
        $request = Tinebase_Http_Request::fromString(
            "POST /index.php HTTP/1.1\r\n" .
            "Host: localhost\r\n" .
            "Content-Type: application/json\r\n" .
            "User-Agent: Mozilla/5.0 (Macintosh) mirall/1.5.0 (build 3709)\r\n"
        );
        // Tinebase_Core::set('request', $request);
        // $response = $this->_execPropfindRequest(expectedStatus: 500)->saveXML();
        Tinebase_Core::set(Tinebase_Core::REQUEST, $request);
        $response = $this->_execPropfindRequest(request: $request, expectedStatus: 500)->saveXML();
        $this->assertStringContainsString(InvalidArgumentException::class, $response);
        $this->assertStringContainsString(sprintf(
            'OwnCloud client min version is "%s"!',
            Tinebase_WebDav_Plugin_OwnCloud::OWNCLOUD_MIN_VERSION
        ), $response);
    }
```

CLI: `root@Tine-dev:/usr/share/tests/tine20# ../../tine20/vendor/bin/phpunit -d memory_limit=-1 -d include_path=/etc/tine20:../../tine20:../../tine20/vendor/zendframework/zendframework1/library/ --configuration phpunit.xml --log-junit ../../tine20/build/reports/junit.xml Tinebase/WebDav/Plugin/OwnCloudTest.php | grep '>>>'`
``` bash
.>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 258ms 0ms - 2024-10-01T18:47:53+00:00 INFO (6): LogListener::startTest::19 [PHPUnit] Starting test: testGetRootsV2
.>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 294ms 0ms - 2024-10-01T18:47:53+00:00 INFO (6): LogListener::startTest::19 [PHPUnit] Starting test: testGetRootsV3
.>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 319ms 0ms - 2024-10-01T18:47:53+00:00 INFO (6): LogListener::startTest::19 [PHPUnit] Starting test: testGetPropertiesV2
.>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 353ms 0ms - 2024-10-01T18:47:53+00:00 INFO (6): LogListener::startTest::19 [PHPUnit] Starting test: testGetPropertiesV3
.>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 396ms 0ms - 2024-10-01T18:47:53+00:00 INFO (6): LogListener::startTest::19 [PHPUnit] Starting test: testGetSizePropertyV2
.>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 421ms 0ms - 2024-10-01T18:47:53+00:00 INFO (6): LogListener::startTest::19 [PHPUnit] Starting test: testGetSizePropertyV3
.>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 451ms 0ms - 2024-10-01T18:47:53+00:00 INFO (6): LogListener::startTest::19 [PHPUnit] Starting test: testGetPropertiesForSharedDirectoryV2
.>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 623ms 0ms - 2024-10-01T18:47:53+00:00 INFO (6): LogListener::startTest::19 [PHPUnit] Starting test: testGetPropertiesForSharedDirectoryV3
.>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 777ms 0ms - 2024-10-01T18:47:53+00:00 INFO (6): LogListener::startTest::19 [PHPUnit] Starting test: testGetPropertiesForSharedDirectoryRights
.>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 901ms 0ms - 2024-10-01T18:47:53+00:00 INFO (6): LogListener::startTest::19 [PHPUnit] Starting test: testInvalidOwnCloudVersion
>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 917ms 0ms - 2024-10-01T18:47:53+00:00 INFO (6): LogListener::startTest::19 [PHPUnit] Starting test: testGetPropertiesWithAccountLoginName
.                                                      12 / 12 (100%)>>>>>>>>>>>User-Agent: Mozilla/5.0 (Macintosh) mirall/2.2.4 (build 3709)<<<<<<<<<a7e3bc - tine20admin 1.04s 2ms - 2024-10-01T18:47:54+00:00 INFO (6): LogListener::endTestSuite::52 [PHPUnit] End test suite: Tinebase_WebDav_Plugin_OwnCloudTest
```

As you can see, the **User Agent is same on all tests** (it should be once '1.5.0' instead of '2.2.4'), hence the old Request is reused. 

The ending of the suite without filter/grep:
```
58c841 - tine20admin 1.19s 1ms - 2024-10-01T18:54:26+00:00 INFO (6): LogListener::endTestSuite::52 [PHPUnit] End test suite: Tinebase_WebDav_Plugin_OwnCloudTest


Time: 1.28 seconds, Memory: 36.00 MB

There was 1 error:

1) Tinebase_WebDav_Plugin_OwnCloudTest::testInvalidOwnCloudVersion
Error: Call to undefined method Tinebase_Http_Request::setBody()

/app/tests/tine20/Tinebase/WebDav/Plugin/OwnCloudTest.php:177
/app/tests/tine20/Tinebase/WebDav/Plugin/OwnCloudTest.php:354

ERRORS!
Tests: 12, Assertions: 43, Errors: 1.
```

I spent really a lot of time on this. You CANNOT change the Request after `::setUp`, it is simply ignored.

paulmhh commented:
thank you very much for the MR, I picked it and changed the test. Also changed some code styles/names. Will let you know once merged

pschuele commented:
I think, we can close the PR, the code has already been merged here: e462559f48384eb3622150af9c8a96ae8e7a6b75
